crates/paimon/src/io/file_io.rs

```diff
@@ -17,66 +17,80 @@
 
 use crate::error::*;
 use std::collections::HashMap;
+use std::ops::Range;
+use std::sync::Arc;
 
-use chrono::offset::Utc;
-use chrono::DateTime;
-use opendal::services::Fs;
-use opendal::{Metakey, Operator};
+use bytes::Bytes;
+use chrono::{DateTime, Utc};
+use opendal::Operator;
 use snafu::ResultExt;
+use url::Url;
+
+use super::Storage;
 
 #[derive(Clone, Debug)]
 pub struct FileIO {
-    op: Operator,
+    inner_storage: Arc<Storage>,
 }
 
 impl FileIO {
-    /// Create a new FileIO.
+    /// Try to infer file io scheme from path.
     ///
     /// The input HashMap is paimon-java's [`Options`](https://github.com/apache/paimon/blob/release-0.8.2/paimon-common/src/main/java/org/apache/paimon/options/Options.java#L60)
-    ///
-    /// TODO: Support building Operator from HashMap via options.
-    pub fn new(_: HashMap<String, String>) -> Result<Self> {
-        let op = Operator::new(Fs::default().root("/"))
-            .context(IoUnexpectedSnafu {
-                message: "Failed to create operator".to_string(),
-            })?
-            .finish();
-        Ok(Self { op })
+    pub fn from_path(path: impl AsRef<str>) -> Result<FileIOBuilder> {
+        let url = Url::parse(path.as_ref())
+            .context(UrlParseSnafu)
+            .or_else(|_| {
+                Url::from_file_path(path.as_ref()).map_err(|_| Error::DataInvalid {
+                    message: "Input is neither a valid URL nor a valid file path".to_string(),
+                })
+            })?;
+
+        Ok(FileIOBuilder::new(url.scheme()))
     }
 
     /// Create a new input file to read data.
     ///
     /// Reference: <https://github.com/apache/paimon/blob/release-0.8.2/paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java#L76>
-    pub fn new_input(&self, path: &str) -> InputFile {
-        InputFile {
-            _op: self.op.clone(),
-            path: path.to_string(),
-        }
+    pub fn new_input(&self, path: impl AsRef<str>) -> crate::Result<InputFile> {
```

Comment:
I think the `impl AsRef<str>` is somewhat overengineered. Accepting a `&str` is sufficient, especially since we're calling `to_string` internally.

Reply:
No problem